portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -15,30 +15,76 @@
 package com.google.sps.servlets;
 
 import com.google.gson.Gson;
+import com.google.sps.models.Comment;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
 import java.io.IOException;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 import java.util.*;
+import java.text.SimpleDateFormat;
 
-/** Servlet that returns some example content. TODO: modify this file to handle comments data */
 @WebServlet("/data")
 public final class DataServlet extends HttpServlet {
 
-  // Messages are hard-coded for testing purposes
-  private ArrayList<String> comments = new ArrayList<String>(
-      Arrays.asList("First comment", "Second comment", "Third comment")
-  );
-
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    Query query = new Query("Comment").addSort("date", SortDirection.DESCENDING);
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+
+    ArrayList<Comment> comments = new ArrayList<>();
+    for (Entity entity: results.asIterable()) {
+      String author = (String) entity.getProperty("author");
+      String email = (String) entity.getProperty("email");
+      String date = (String) entity.getProperty("date");
+      String commentText = (String) entity.getProperty("comment");
+
+      Comment comment = new Comment(author, email, date, commentText);
+      comments.add(comment);
+    }
+
     String json = convertArrayListToJson(comments);
     response.setContentType("application/json");
     response.getWriter().println(json);
   }
 
-  private String convertArrayListToJson(ArrayList<String> comments) {
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+
+    String author = request.getParameter("comment_author").trim();
+    if (author.isEmpty()) author = "Anonymous";
+
+    String email = request.getParameter("email").trim();
+    if (email.isEmpty()) email = "@";
+
+    String comment = request.getParameter("comment").trim();
```

Comment:
You should probably do this upfront because there's no point getting author and email if you're going to stop anyway

Reply:
Also perhaps you'll get to this later in the training, so no need to necessarily do anything about this now, but just wanted to point out this would be a good example for when to start using your UX thinking!  Like, you could think, instead of just returning and leaving it unclear to the user whether clicking the submit button did what they expected it to (i.e. submitted the comment), you could flash a dialog or something to indicate an error state.

Reply:
Woah that's a really good point! Yeah I was definitely thinking to myself that the way this is set up now, if a comment got posted but didn't go through, the user wouldn't really have any idea what happened. I think the validation for required fields on the frontend should be enough to prevent something weird from being submitted, but it would make sense that in the off chance that something weird did make it to the back, the server would throw an error that would alert the user of what went wrong. I'll look into it.